File: 2019/05/intcode.py
#!/usr/bin/env python3
class IntcodeInterpreter:
    def __init__(self, program):
        assert(all(isinstance(elem, int) for elem in program))
        self.program = program[:]
        self.pc = 0

    def consume(self, count):
        vals = self.program[self.pc:self.pc + count]
        self.pc += count
        return vals

    def consumePtrs(self, count, modes):
        return [addr if modes // 10**ind % 10 == 1 else self.program[addr]
                for ind, addr in enumerate(self.consume(count))]

    def writePtr(self, val):
        self.program[self.consume(1)[0]] = val

    def step(self):
        modes, opcode = divmod(self.consume(1)[0], 100)
        if opcode == 1:
            self.writePtr(sum(self.consumePtrs(2, modes)))
        elif opcode == 2:
            val1, val2 = self.consumePtrs(2, modes)
            self.writePtr(val1 * val2)
        elif opcode == 3:
            self.writePtr(int(input("Need input: ")))
        elif opcode == 4:
            print(self.consumePtrs(1, modes))
        elif opcode == 99:
            self.pc = None
        else:
            raise Exception(f"Bad opcode {opcode}")

    def run(self):
        while self.pc is not None:
            self.step()


def execute_loop(program, args):
    interp = IntcodeInterpreter(program)
    interp.program[1:3] = args
    interp.run()

    return interp.program[0]


def main():
    program = open("input").read().replace('\n', '').split(',')
    program = [int(x) for x in program]

    result = execute_loop(program, [12, 2])
    print(result)

    for noun in range(0, len(program)):
        for verb in range(0, len(program)):
            result = execute_loop(program, [noun, verb])
            if result == 19690720:
                print(noun * 100 + verb)
                break


main()

File: 2019/05/intcode.scala
#!/usr/bin/env scala
import Integral.Implicits._

class IntcodeInterpreter(origProgram: Seq[Int]) {
  var pc = 0
  var program = origProgram.toBuffer

  sealed trait Opcode {
    def run(modes: Int): Unit
  }
  private case class UnitOpcode(args: Int, func: Seq[Int] => Unit) extends Opcode {
    def run(modes: Int) = func(consumePtrs(args, modes))
  }
  private case class WritebackOpcode(args: Int, func: Seq[Int] => Int) extends Opcode {
    def run(modes: Int) = writePtr(func(consumePtrs(args, modes)))
  }

  val opcodes = Map[Int, Opcode](
    1 -> WritebackOpcode(2, _.sum),
    2 -> WritebackOpcode(2, _.product),
    3 -> WritebackOpcode(0, { _ => io.StdIn.readLine("Need input: ").toInt}),
    4 -> UnitOpcode(1, println),
    5 -> UnitOpcode(2, { p => if (p(0) != 0) pc = p(1) }),
    6 -> UnitOpcode(2, { p => if (p(0) == 0) pc = p(1) }),
    7 -> WritebackOpcode(2, { p => if (p(0) < p(1)) 1 else 0 }),
    8 -> WritebackOpcode(2, { p => if (p(0) == p(1)) 1 else 0 }),
    99 -> UnitOpcode(0, { _ => pc = -1 }),
  )

  def consume(count: Int): Seq[Int] = {
    val res = program.slice(pc, pc + count)
    pc += count
    res.toSeq
  }

  def consumePtrs(count: Int, modes: Int): Seq[Int] = {
    def alterPtrs(ptrs: Seq[Int], modes: Int): List[Int] =
      if (ptrs.isEmpty) Nil
      else if (modes % 10 == 1) ptrs.head :: alterPtrs(ptrs.tail, modes / 10)
      else program(ptrs.head) :: alterPtrs(ptrs.tail, modes / 10)
    alterPtrs(consume(count), modes)
  }

  def writePtr(value: Int): Unit = program(consume(1)(0)) = value

  def step(): Unit = {
    val (modes, opcode) = consume(1)(0) /% 100
    opcodes.getOrElse(opcode, throw new Exception(s"Bad opcode $opcode"))
      .run(modes)
  }

  def run(): Unit = while (pc >= 0) step()
}

def main(): Unit = {
  val program = io.Source.fromFile("input").getLines
    .flatMap(_.split(",")).map(_.toInt).toSeq
  new IntcodeInterpreter(program).run()
}

main()
